File: scripts/validate-graph.mjs
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = fileURLToPath(new URL('..', import.meta.url));
const readJson = (path) => JSON.parse(readFileSync(join(root, path), 'utf8'));
const fail = (message) => { throw new Error(message); };
const ensure = (condition, message) => { if (!condition) fail(message); };

const index = readJson('graph/index.json');
for (const path of [...Object.values(index.canonical_files), ...Object.values(index.schemas)]) {
  ensure(typeof path === 'string' && path.length > 0, 'Invalid canonical/schema path');
  ensure(existsSync(join(root, path)), `Missing declared graph file: ${path}`);
}
const nodes = readJson(index.canonical_files.nodes);
const edges = readJson(index.canonical_files.edges);
const moves = readJson(index.canonical_files.research_moves);
const collections = readJson(index.canonical_files.collections);

for (const schemaPath of Object.values(index.schemas)) readJson(schemaPath);

const nodeIds = new Set();
const edgeIds = new Set();
const moveIds = new Set();
const collectionIds = new Set();
const kinds = new Set(['domain', 'bridge', 'problem']);
const confidences = new Set(['formal', 'literature', 'heuristic', 'speculative']);
const relations = new Set(['depends_on', 'generalizes', 'specializes', 'analogy', 'dual', 'formalizes_as', 'transfers_via', 'obstructs', 'suggests', 'tests', 'uses', 'bridge']);

for (const node of nodes) {
  ensure(/^((domain|bridge|problem)\.)[a-z0-9_\.]+$/.test(node.id), `Invalid node id: ${node.id}`);
  ensure(!nodeIds.has(node.id), `Duplicate node id: ${node.id}`);
  ensure(kinds.has(node.kind), `${node.id}: invalid kind ${node.kind}`);
  ensure(confidences.has(node.confidence), `${node.id}: invalid confidence ${node.confidence}`);
  ensure(node.id.startsWith(`${node.kind}.`), `${node.id}: id prefix does not match kind`);
  ensure(typeof node.title === 'string' && node.title.trim().length >= 3, `${node.id}: missing title`);
  ensure(typeof node.summary === 'string' && node.summary.trim().length >= 20, `${node.id}: summary is too short`);
  ensure(Array.isArray(node.tags) && node.tags.length >= 2, `${node.id}: needs at least two tags`);
  ensure(new Set(node.tags).size === node.tags.length, `${node.id}: duplicate tags`);
  ensure(Array.isArray(node.questions) && node.questions.length >= 1, `${node.id}: needs at least one live question`);
  ensure(node.lean && Array.isArray(node.lean.imports) && Array.isArray(node.lean.declarations) && Array.isArray(node.lean.targets), `${node.id}: invalid Lean metadata`);
  ensure(node.lean.targets.length >= 1, `${node.id}: needs a Lean target`);
  if (node.kind === 'problem') ensure(['solved', 'unsolved'].includes(node.status), `${node.id}: problem status required`);
  else ensure(node.status === undefined, `${node.id}: status is only valid for problems`);
  if (node.references) {
    for (const reference of node.references) {
      ensure(typeof reference.label === 'string' && reference.label.trim(), `${node.id}: reference label required`);
      ensure(/^https:\/\//.test(reference.url), `${node.id}: references must use HTTPS`);
      ensure(['official', 'paper', 'book', 'documentation', 'survey'].includes(reference.type), `${node.id}: invalid reference type`);
    }
  }
  nodeIds.add(node.id);
}

const degree = new Map(nodes.map((node) => [node.id, 0]));
const semanticEdges = new Set();
for (const edge of edges) {
  ensure(/^edge\.[a-z0-9_\.]+$/.test(edge.id), `Invalid edge id: ${edge.id}`);
  ensure(!edgeIds.has(edge.id), `Duplicate edge id: ${edge.id}`);
  ensure(nodeIds.has(edge.source), `${edge.id}: missing source ${edge.source}`);
  ensure(nodeIds.has(edge.target), `${edge.id}: missing target ${edge.target}`);
  ensure(edge.source !== edge.target, `${edge.id}: self-edge`);
  ensure(relations.has(edge.relation), `${edge.id}: invalid relation ${edge.relation}`);
  ensure(confidences.has(edge.confidence), `${edge.id}: invalid confidence ${edge.confidence}`);
  ensure(typeof edge.mechanism === 'string' && edge.mechanism.trim().length >= 20, `${edge.id}: mechanism is too short`);
  const semanticKey = `${edge.source}|${edge.relation}|${edge.target}`;
  ensure(!semanticEdges.has(semanticKey), `${edge.id}: duplicate source/relation/target claim`);
  semanticEdges.add(semanticKey);
  if (edge.confidence === 'speculative') ensure(typeof edge.notes === 'string' && /falsif/i.test(edge.notes), `${edge.id}: speculative edges need a falsifier note`);
  degree.set(edge.source, degree.get(edge.source) + 1);
  degree.set(edge.target, degree.get(edge.target) + 1);
  edgeIds.add(edge.id);
}

for (const [nodeId, count] of degree) ensure(count > 0, `${nodeId}: isolated node`);

for (const move of moves) {
  ensure(/^move\.[a-z0-9_\.]+$/.test(move.id), `Invalid move id: ${move.id}`);
  ensure(!moveIds.has(move.id), `Duplicate move id: ${move.id}`);
  ensure(typeof move.title === 'string' && move.title.trim().length >= 2, `${move.id}: title required`);
  ensure(typeof move.description === 'string' && move.description.length >= 20, `${move.id}: description is too short`);
  ensure(Array.isArray(move.good_for) && move.good_for.length >= 1, `${move.id}: good_for required`);
  for (const nodeId of move.good_for) ensure(nodeIds.has(nodeId), `${move.id}: unknown good_for node ${nodeId}`);
  ensure(typeof move.output === 'string' && move.output.trim().length >= 8, `${move.id}: output required`);
  ensure(Array.isArray(move.risks) && move.risks.length >= 1, `${move.id}: risks required`);
  ensure(typeof move.lean_test === 'string' && move.lean_test.length >= 12, `${move.id}: Lean test required`);
  moveIds.add(move.id);
}

for (const collection of collections) {
  ensure(/^collection\.[a-z0-9_\.]+$/.test(collection.id), `Invalid collection id: ${collection.id}`);
  ensure(!collectionIds.has(collection.id), `Duplicate collection id: ${collection.id}`);
  ensure(typeof collection.title === 'string' && collection.title.trim().length >= 2, `${collection.id}: title required`);
  ensure(typeof collection.description === 'string' && collection.description.trim().length >= 12, `${collection.id}: description required`);
  ensure(Array.isArray(collection.nodes) && collection.nodes.length >= 2, `${collection.id}: needs at least two nodes`);
  ensure(new Set(collection.nodes).size === collection.nodes.length, `${collection.id}: duplicate nodes`);
  for (const nodeId of collection.nodes) ensure(nodeIds.has(nodeId), `${collection.id}: unknown node ${nodeId}`);
  collectionIds.add(collection.id);
}

for (const rootNode of index.root_nodes) ensure(nodeIds.has(rootNode), `Unknown root node: ${rootNode}`);
for (const entrypoint of index.agent_entrypoints) {
  ensure(typeof entrypoint === 'string' && entrypoint.length > 0, 'Invalid agent entrypoint');
  ensure(existsSync(join(root, entrypoint)), `Missing agent entrypoint: ${entrypoint}`);
}

const byKind = Object.fromEntries([...kinds].sort().map((kind) => [kind, nodes.filter((node) => node.kind === kind).length]));
const byConfidence = Object.fromEntries([...confidences].sort().map((confidence) => [confidence, nodes.filter((node) => node.confidence === confidence).length]).filter(([, count]) => count > 0));
ensure(index.stats.nodes === nodes.length, 'index.stats.nodes is stale');
ensure(index.stats.edges === edges.length, 'index.stats.edges is stale');
ensure(index.stats.research_moves === moves.length, 'index.stats.research_moves is stale');
ensure(index.stats.collections === collections.length, 'index.stats.collections is stale');
ensure(JSON.stringify(index.stats.by_kind) === JSON.stringify(byKind), 'index.stats.by_kind is stale');
ensure(JSON.stringify(index.stats.by_confidence) === JSON.stringify(byConfidence), 'index.stats.by_confidence is stale');

console.log(`Validated canonical graph: ${nodes.length} nodes, ${edges.length} edges, ${moves.length} research moves, ${collections.length} collections.`);

File: scripts/validate_graph.py
from __future__ import annotations

import json
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]

CONFIDENCE = {"formal", "literature", "heuristic", "speculative"}
NODE_KINDS = {"domain", "theory", "concept", "theorem", "formula", "method", "problem", "bridge"}
RELATIONS = {
    "depends_on",
    "generalizes",
    "specializes",
    "analogy",
    "dual",
    "formalizes_as",
    "transfers_via",
    "obstructs",
    "suggests",
    "tests",
    "uses",
    "bridge",
}


def load_json(path: Path):
    with path.open("r", encoding="utf-8") as handle:
        return json.load(handle)


def fail(message: str) -> None:
    print(f"ERROR: {message}", file=sys.stderr)
    raise SystemExit(1)


def require_fields(item: dict, fields: set[str], label: str) -> None:
    missing = sorted(fields - set(item))
    if missing:
        fail(f"{label} missing fields: {', '.join(missing)}")


def main() -> None:
    nodes = load_json(ROOT / "graph" / "nodes" / "core.json")
    edges = load_json(ROOT / "graph" / "edges.json")
    moves = load_json(ROOT / "graph" / "research_moves.json")

    node_fields = {"id", "kind", "title", "summary", "confidence", "tags", "lean", "questions"}
    edge_fields = {"id", "source", "target", "relation", "confidence", "mechanism"}
    move_fields = {"id", "title", "description", "good_for", "output", "risks", "lean_test"}

    seen_nodes: set[str] = set()
    for node in nodes:
        require_fields(node, node_fields, f"node {node.get('id', '<unknown>')}")
        node_id = node["id"]
        if node_id in seen_nodes:
            fail(f"duplicate node id: {node_id}")
        seen_nodes.add(node_id)
        if node["kind"] not in NODE_KINDS:
            fail(f"node {node_id} has invalid kind {node['kind']}")
        if node["confidence"] not in CONFIDENCE:
            fail(f"node {node_id} has invalid confidence {node['confidence']}")
        lean = node["lean"]
        require_fields(lean, {"imports", "declarations", "targets"}, f"node {node_id}.lean")
        for key in ("tags", "questions"):
            if not isinstance(node[key], list):
                fail(f"node {node_id}.{key} must be a list")

    seen_edges: set[str] = set()
    for edge in edges:
        require_fields(edge, edge_fields, f"edge {edge.get('id', '<unknown>')}")
        edge_id = edge["id"]
        if edge_id in seen_edges:
            fail(f"duplicate edge id: {edge_id}")
        seen_edges.add(edge_id)
        if edge["source"] not in seen_nodes:
            fail(f"edge {edge_id} source does not exist: {edge['source']}")
        if edge["target"] not in seen_nodes:
            fail(f"edge {edge_id} target does not exist: {edge['target']}")
        if edge["relation"] not in RELATIONS:
            fail(f"edge {edge_id} has invalid relation {edge['relation']}")
        if edge["confidence"] not in CONFIDENCE:
            fail(f"edge {edge_id} has invalid confidence {edge['confidence']}")

    seen_moves: set[str] = set()
    for move in moves:
        require_fields(move, move_fields, f"move {move.get('id', '<unknown>')}")
        move_id = move["id"]
        if move_id in seen_moves:
            fail(f"duplicate move id: {move_id}")
        seen_moves.add(move_id)
        for node_id in move["good_for"]:
            if node_id not in seen_nodes:
                fail(f"move {move_id} references missing node {node_id}")

    print(f"OK: {len(nodes)} nodes, {len(edges)} edges, {len(moves)} research moves.")


if __name__ == "__main__":
    main()

